test/distributed/org/apache/cassandra/distributed/upgrade/CompressionUpgradeTest.java

```diff
@@ -0,0 +1,105 @@
+    {
+        String[][] zstdQueries = forZstd();
+        run(new String[][]{
+        zstdQueries[0], // min
+        zstdQueries[1], // default
+        zstdQueries[2], // max
+        forCompressor(SnappyCompressor.class),
+        forCompressor(DeflateCompressor.class),
+        forCompressor(LZ4Compressor.class)
+        });
+    }
+
+    private String[][] forZstd()
+    {
+        String tableTemplate = "CREATE TABLE %s.%s (pk int, ck int, v int, PRIMARY KEY (pk, ck)) WITH compression = {'class': '" + ZstdCompressor.class.getSimpleName() + "', 'compression_level': %s}";
+        String insertTemplate = "INSERT INTO %s.%s (pk, ck, v) VALUES (1, 1, 1)";
+        String selectTemplate = "SELECT * FROM %s.%s WHERE pk = 1";
+
+        // Zstd min and max compression levels, (Zstd.{min,default,max}CompressionLevel())
+        // It should be possible to create a table with this level in each version and read / write data with it
+        String[][] params = new String[][]{ { "tbl_zstd_min_compression_level", Integer.toString(-131072) },
+                                            { "tbl_zstd_max_compression_level", Integer.toString(22) },
+                                            { "tbl_zstd_default_compression_level", Integer.toString(3) } };
+
+        String[][] zstdQueries = new String[params.length][params.length];
+
+        for (int i = 0; i < params.length; i++)
+        {
+            zstdQueries[i][0] = format(tableTemplate, KEYSPACE, params[i][0], params[i][1]);
+            zstdQueries[i][1] = format(insertTemplate, KEYSPACE, params[i][0]);
+            zstdQueries[i][2] = format(selectTemplate, KEYSPACE, params[i][0]);
+        }
+
+        return zstdQueries;
+    }
+
+    private String[] forCompressor(Class<?> compressorClass)
+    {
+        String compressorName = compressorClass.getSimpleName();
+        return new String[]{
+        "CREATE TABLE " + KEYSPACE + ".tbl_" + compressorName + " (pk int, ck int, v int, PRIMARY KEY (pk, ck)) WITH compression = {'class': '" + compressorName + "'}",
+        "INSERT INTO " + KEYSPACE + ".tbl_" + compressorName + " (pk, ck, v) VALUES (1, 1, 1)",
+        "SELECT * FROM " + KEYSPACE + ".tbl_" + compressorName + " WHERE pk = 1"
+        };
+    }
+
+    private void run(String[][] queries) throws Throwable
+    {
+        new TestCase()
+        .nodes(2)
+        .nodesToUpgrade(1)
+        .withConfig((cfg) -> cfg.with(NETWORK, GOSSIP))
+        .upgradesToCurrentFrom(OLDEST)
+        .setup((cluster) -> {
+            for (int i = 0; i < queries.length; i++)
+            {
+                cluster.schemaChange(queries[i][0]);
+                cluster.coordinator(1).execute(queries[i][1], ALL);
+            }
+        })
+        .runAfterNodeUpgrade((cluster, node) -> {
+            for (int i : new int[]{ 1, 2 })
+                for (int j = 0; j < queries.length; j++)
+                    assertRows(cluster.coordinator(i).execute(queries[j][2], ALL), row(1, 1, 1));
```

Comment:
Thanks for adding the test!

I have a _nit_ on providing a better insight on test failures. Can you include the compressor and cassandra version in the assertion failure? The test iterates through bunch of configurations, it would be hard to see which configuration fails.

Reply:
The `assertRows` is an utility provided from dtest-api. You might need to catch the `AssertionError` in order to enrich the message...

Reply:
done